File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.js
function solution(friends, gifts) {
    const giftMap = friends.reduce((acc, giver, i) => {
        acc[giver] = {}
        friends.forEach((receiver, j) => {
            if (i !== j) acc[giver][receiver] = 0
        })
        return acc
    }, {})

    const giftIndex = friends.reduce((acc, name) => {
        acc[name] = 0
        return acc
    }, {})
    
    gifts.forEach((gift) => {
        const [giver, receiver] = gift.split(' ')
        giftMap[giver][receiver]++
        giftIndex[giver]++
        giftIndex[receiver]--
    })
    console.log(giftMap)
    // {
    //   muzi: { ryan: 0, frodo: 2, neo: 0 },
    //   ryan: { muzi: 3, frodo: 0, neo: 0 },
    //   frodo: { muzi: 1, ryan: 1, neo: 0 },
    //   neo: { muzi: 1, ryan: 0, frodo: 0 }
    // }
    
    console.log('giftIndex', giftIndex)
    // giftIndex { muzi: -3, ryan: 2, frodo: 0, neo: 1 }
    
    const nextMonth = friends.reduce((acc, name) => {
        acc[name] = 0
        return acc
    }, {})
    // { muzi: 0, ryan: 0, frodo: 0, neo: 0 }
    
    Object.entries(giftMap).forEach(([giver, receiversObj]) => { // muzi { ryan: 0, frodo: 2, neo: 0 }
        // 두 사람 중 더 많은 선물 준 사람++
        Object.entries(receiversObj).forEach(([receiver, giveCount]) => {
            const receiveCount = giftMap[receiver][giver]
            if (giveCount > receiveCount) nextMonth[giver]++
            else if (giveCount < receiveCount) nextMonth[receiver]++
            else {
                // 선물 지수 높은 사람++
                if (giftIndex[giver] > giftIndex[receiver]) nextMonth[giver]++
                else if (giftIndex[giver] < giftIndex[receiver]) nextMonth[receiver]++
                else return
            }
        })
    })
    
    return Math.max(...Object.values(nextMonth))/2
} 

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.py
def solution(friends, gifts):
    gift_dict = {}
    
    for name in friends:
        gift_dict[name] = {}
        for friend in friends:
            if name != friend:
                gift_dict[name][friend] = 0
    
    # 선물 주고받기 기록
    for gift in gifts:
        giver, receiver = gift.split(' ')
        gift_dict[giver][receiver] += 1
        
    # 각 사람이 다음 달에 받을 선물 개수를 저장할 딕셔너리
    next_month_gifts = {name: 0 for name in gift_dict}

    # 각 사람을 순회하며 선물 계산
    for giver, receivers in gift_dict.items():
        for receiver, give_count in receivers.items():
            receive_count = gift_dict[receiver][giver]
            
            # 더 많이 준 사람이 다음 달에 하나 받음
            if give_count > receive_count: 
                next_month_gifts[giver] += 1
            elif give_count < receive_count:
                next_month_gifts[receiver] += 1
            # 선물 지수로 결정 (선물 지수 : 내가 준 선물 수 - 내가 받은 선물 수 )
            else: 
                giver_gift_index = sum(gift_dict[giver].values()) - sum(gift_dict[name][giver] for name in gift_dict if name != giver)
                receiver_gift_index = sum(gift_dict[receiver].values()) - sum(gift_dict[name][receiver] for name in gift_dict if name != receiver)
                if giver_gift_index > receiver_gift_index:
                    next_month_gifts[giver] += 1
                elif giver_gift_index < receiver_gift_index:
                    next_month_gifts[receiver] += 1
                else:
                    continue
        
    return max(next_month_gifts.values()) / 2
